### DISCUSSION TITLE
WIFI, MQTT, REST API added to Sensorbox-2 -- PLEASE TEST

### DISCUSSION THREAD

**Discussion by dingo35** [dingo35/Sensorbox-2#1] [Announcements]
I am looking for Sensorbox-2 users who are willing and able to test my new firmware version.

The new firmware version now also supports WiFi, MQTT, REST API, automatic time zone configuration, in fact pretty much everything I could take over from the SmartEVSE is now included. In addition, the "takeover" is designed in such a way that large parts of the code will be shared with the SmartEVSE, so that maintenance remains bearable.

This means that the Sensorbox-2 can also be used without a modbus connection to the SmartEVSE.

I only have one major problem: at my current location I do not have a 3x230V 50Hz grid available, which means I cannot test the core functionality.

[3a59ad9c04.zip](https://github.com/user-attachments/files/17245641/3a59ad9c04.zip)

(You can flash easily by going to http://ipaddress-of-your-sensorbox/update).

To configure wifi please look at the docs:
https://github.com/dingo35/Sensorbox-2/blob/master/docs/configuration.md

**Comment by dingo35** [comment]
Additional info:
Backup your original firmware:
esptool.py read_flash 0x00000 0x400000 backup.bin

Flash my firmware.bin:
esptool.py --port /dev/ttyUSB0 -b 115200 write_flash 0x10000 firmware.bin

and again on:
esptool.py --port /dev/ttyUSB0 -b 115200 write_flash 0x1C0000 firmware.bin

so both OTA0 and OTA1 parititons have the same version.

Restore your backup:

Flash your backup:
esptool.py write_flash 0x00000 backup.bin

**Comment by dingo35** [comment]
I solved two bugs:
-the backup wifi login procedure via USB command line did not work, fixed now. ESPTouch always worked....
-currents were reported 10x too low ...

Fixed:
[613aad7aa.zip](https://github.com/user-attachments/files/17262606/613aad7aa.zip)

**Comment by Bl4ckr4v3n** [comment]
I flashed 613aad7aa on both partitions through USB, but I have no idea whether is was successful, is there a way to check the current firmware version?
FYI: I connected the unit back to the smartEVSE and it looks to be functioning (car not yet connected however), no errors and P1 data is visible on the display.
However I'd expect to see the additional WiFi functionality and I don't see any, so it feels to me as if the original firmware is still there.

**Comment by dingo35** [comment]
You should see a new device asking for a DHCP lease on your Access Point; if you surf to that ip address it shows you a screen with a.o. serialnr and firmware version.

You should also see it present on your mDNS ; on linux do "avahi-browse -a"

**Comment by dingo35** [comment]
By the way, did you follow the procedure to log your Sensorbox onto your wifi?

**Comment by Bl4ckr4v3n** [comment]
Well, the procedure implies you have a WiFi option on the evse screen, which is not there on the smartEVSE v2. I don't know if the WiFi is disabled by default, but otherwise I'd need to switch it on through serial or the like.

**Comment by dingo35** [comment]
No, you can activate wifi on the Sensorbox-2 via the ESPtouch procedure or the USB procedure.

Especially if your SmartEVSEv2 does not support wifi, it will bring you:
1. More options on your sensorbox webserver screen
2. MQTT feed of your Sensorbox currents to your MQTT broker
3. Easy firmware update of your Sensorbox via /update
4. REST API access to your Sensorbox currents

**Reply by Bl4ckr4v3n** [reply]
I tried all day to figure this out, but I guess I look in the wrong places. It should go in some kind of discovery mode AFAIK, but I can't detect or trigger this. 
All SmartEVSE manuals point to the display to switch it on, both for the ESPtouch and USB procedures. But with a SmartEVSE V2 without the Wifi menu it's not possible to activate the smart config function from there.
It still looks as if the wifi is completely down, I don't detect any client or AP activity with my network monitors.

**Comment by dingo35** [comment]
We are talking SENSORBOX  here, not SmartEVSE.
As stated, here is the procedure:
To configure wifi please look at the docs:
https://github.com/dingo35/Sensorbox-2/blob/master/docs/configuration.md

**Reply by Bl4ckr4v3n** [reply]
I overlooked that one, sorry...
Nevertheless USB procedure should be the same, but my sensorbox doesn't stop spitting the serial telegrams after hitting enter.
Now I tried the ESPtouch procedure also, but that doesn't give anything as well unfortunately.
It still feels like it didn't take the new firmware, although it flashed both partitions without errors.
Would it be possible that the ESP reverts the old partition to the newly flashed partition during the reboots?
(I presume that the message 'hard resetting via RTS pin' at the end of every flash means it's rebooting)
i.e. is there command to flash both partitions in one run to make sure it doesn't fallback to the old firmware?

**Reply by Bl4ckr4v3n** [reply]
I combined the commands like this:
esptool.py --port /dev/ttyUSB0 -b 115200 write_flash 0x10000 firmware.bin 0x1C0000 firmware.bin
Unfortunately still the same result, no connection on the ESPtouch app and in the serial connection it keeps spitting telegrams which 'Enter' doesn't stop.

**Comment by dingo35** [comment]
Ok thx for your efforts, Im working on some modifications so Ill be back here!

**Reply by Bl4ckr4v3n** [reply]
Any update on the modifications in the last month?

**Comment by dingo35** [comment]
You might have noticed we released v2.1.0 ; thanks for your efforts!

**Reply by the-master006** [reply]
Thanks for version 2.1.0. The MQTT works perfectly!! 
1 Question although, I see it is only publishing the currents on the 3 phases and no voltages or cumulative power, power factor (not sure if that is in a P1 meter actually...). Can you alsu publish all other values? Thanks! 
<img width="397" height="258" alt="image" src="https://github.com/user-attachments/assets/e5af4016-607b-4407-84c4-5f99de090960" />
This is all what the MQTT function sends

**Comment by dingo35** [comment]
On p1 telegrams the voltage is optional, it isnt of much value since its always around 230V. PF is not in there.

**Reply by the-master006** [reply]
hmm ok thanks, clear

**Reply by the-master006** [reply]
I don't agree there by the way, I tracked the voltages today via my P1 USB DSMR meter, and you see here the phases as measured by the DSMR 5.0 ISKRA meter with the P1 USB kabel. They are pretty accurate. I also compared this with my PV measurements and there is quite a good match. So is there any possibility to add these voltages and possibly other parameters to the sensorbox MQTT output? Thanks!
<img width="770" height="416" alt="image" src="https://github.com/user-attachments/assets/fa115e10-7dad-4261-bf75-96be2d842511" />

**Comment by dingo35** [comment]
You are in luck, the P1 routine seems to detect the voltages per phase. I don't have a DSMR meter here so all the testing has to be done extensively by you; not only checking the values + their accuracy, but also if they are correctly announced in Home Assistant.

I need you to test really good, since I don't want to bring out a crappy version of the Sensorbox firmware ....

[f826636ba.zip](https://github.com/user-attachments/files/21768179/f826636ba.zip)

**Reply by the-master006** [reply]
I flashed the firmware, and the MQTT output works perfectly: 
<img width="248" height="208" alt="image" src="https://github.com/user-attachments/assets/46c65e46-89f9-4010-9f16-2b601e01f7fb" />
What I do see is that the P1 meter has output with 1 decimal, while the sensorbox shows now no decimals.
What do you mean with correct announcing in home assistant? I am reading only the Mqtt feed, I am not reading the sensorbox in homeassistant (did not know there is a plugin or so for that?)

**Comment by dingo35** [comment]
If you have MQTT configured in HA, it will discover the Sensorbox outputs automatically, with the right unit (e.g. Ampéres or Volts) and multiplier (Volts, milliVolts etc.).

The Volts are read into a float, its easy to publish in deciVolts but that would generate a lot of confusion. HA has poor documentation so no clue how to publish with one decimal.

**Reply by the-master006** [reply]
Ok thanks. Anyway, it seems the registration and display in Homeassistant also work fine: 
<img width="547" height="665" alt="image" src="https://github.com/user-attachments/assets/3e13fa31-71b6-40c2-9c8e-cc7a45b270ad" />

**Comment by dingo35** [comment]
I found a way, outside of HA, to publish the values as a string, which would make publishing with one decimal easy.
I have some other stuff going on, I'll look at it somewhere next week.

**Reply by the-master006** [reply]
Thanks! I appreciate your help and efforts!

**Comment by dingo35** [comment]
This version should publish with 1 decimal accuracy, especially anxious to see how this is interpreted in Home Assistant, please test:
[41a741f9.zip](https://github.com/user-attachments/files/21829204/41a741f9.zip)

**Reply by the-master006** [reply]
Seems to work perfectly! 
<img width="1170" height="670" alt="image" src="https://github.com/user-attachments/assets/4085931d-c693-47d3-b4ca-5265a550779d" />

**Comment by dingo35** [comment]
I dont see the 1 decimal accuracy?!

**Reply by the-master006** [reply]
yes sorry, you are right, I mixed it up with the amp's. Voltage is indeed not showing 1 digit. in mqtt explorer however it is: 
<img width="203" height="167" alt="image" src="https://github.com/user-attachments/assets/9a13590c-1f4f-4b0c-9b3f-04cfa2f1588c" />

**Comment by dingo35** [comment]
Could you check if this version gives 1 decimal in HomeAssistant?
[b5081364b.zip](https://github.com/user-attachments/files/21989634/b5081364b.zip)

**Reply by the-master006** [reply]
This is an interesting one: in the history of the Entity it shows now the voltage with at least 1 digit: See the graph when from 21.11h I installed your latest FW version, while before in only made increments in whole Volts.
<img width="839" height="254" alt="image" src="https://github.com/user-attachments/assets/4cf0eaa2-a161-44cf-bbd6-aa888f6b2e8d" />
But under the devices under MQTT I still see the Volts without digit: 
<img width="643" height="414" alt="image" src="https://github.com/user-attachments/assets/e4f3da7f-1f8c-4ff1-9acc-1b1443d9956f" />

**Comment by dingo35** [comment]
How about when you remove the device from HA and let it be re-discovered?

**Reply by the-master006** [reply]
Tried that, but no difference
